### PR TITLE
HDDS-7690. [Snapshot] Use SST file list output from compaction DAG as SnapshotDiff input

### PR DESCRIPTION
**Pull request by smengcl**
## What changes were proposed in this pull request?

This is a follow-up to [HDDS-7466](https://issues.apache.org/jira/browse/HDDS-7466).

Currently [HDDS-7466](https://issues.apache.org/jira/browse/HDDS-7466) is not being passed in the "filtered" SST list result from compaction DAG output.

The original jira HDDS-7466 does not have UTs. Some test are being added in [HDDS-7607](https://issues.apache.org/jira/browse/HDDS-7607). More tests to come in [HDDS-7467](https://issues.apache.org/jira/browse/HDDS-7467).

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7690

## How was this patch tested?

- [x] Integration test from #4108 passed.
- [x] Manually tested.

### CODE REVIEW

**Review by prashantpogde — APPROVED** [review]
Changes look good to me. Thanks @smengcl

**Comment by prashantpogde** [comment]
Thanks for updating @smengcl . Looks good.

**Review by prashantpogde — APPROVED** [review]

**Comment by smengcl** [comment]
Thanks @prashantpogde for reviewing this.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java [hunk 1]
```diff
@@ -44,24 +50,74 @@
...
+    this.differ = differ;
+  }
+
+  /**
+   * Copied straight from TestOMSnapshotDAG. TODO: Dedup. Move this to util.
+   */
+  private Map<String, String> getTablePrefixes(
+          OMMetadataManager omMetadataManager,
+          String volumeName, String bucketName) throws IOException {
+    HashMap<String, String> tablePrefixes = new HashMap<>();
```
Comment (hemantk-12) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:74]:
```suggestion
    Map<String, String> tablePrefixes = new HashMap<>();
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java [hunk 2]
```diff
@@ -44,24 +50,74 @@
...
+  private static final Logger LOG =
+          LoggerFactory.getLogger(SnapshotDiffManager.class);
+  private RocksDBCheckpointDiffer differ;
+
+  public SnapshotDiffManager(RocksDBCheckpointDiffer differ) {
+    this.differ = differ;
+  }
+
+  /**
+   * Copied straight from TestOMSnapshotDAG. TODO: Dedup. Move this to util.
```
Comment (hemantk-12) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:69]: This should be just a comment not java-doc comment.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java [hunk 3]
```diff
@@ -44,24 +50,74 @@
...
+                                          final String bucketName)
+          throws IOException {
+
+    final OMMetadataManager snapshotOMMM = omSnapshot.getMetadataManager();
+    final String checkpointPath =
+            snapshotOMMM.getStore().getDbLocation().getPath();
+    final String snapshotId = snapshotInfo.getSnapshotID();
+    final long dbTxSequenceNumber = snapshotInfo.getDbTxSequenceNumber();
+
+    DifferSnapshotInfo dsi = new DifferSnapshotInfo(
```
Comment (hemantk-12) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:102]:
```suggestion
    return new DifferSnapshotInfo(
            checkpointPath,
            snapshotId,
            dbTxSequenceNumber,
            getTablePrefixes(snapshotOMMM, volumeName, bucketName));
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java [hunk 4]
```diff
@@ -152,19 +213,67 @@ private String getKeyOrDirectoryName(boolean isDirectory,
...
+      // Construct DifferSnapshotInfo
+      final DifferSnapshotInfo fromDSI =
+              getDSIFromSI(fsInfo, fromSnapshot, volume, bucket);
+      final DifferSnapshotInfo toDSI =
+              getDSIFromSI(tsInfo, toSnapshot, volume, bucket);
+
+      LOG.debug("Calling RocksDBCheckpointDiffer");
+      List<String> sstDiffList =
+              differ.getSSTDiffListWithFullPath(toDSI, fromDSI);
+      LOG.debug("SST diff list: {}", sstDiffList);
```
Comment (hemantk-12) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:237]: nit: This is duplicate. It gets logged in RocksDBCheckpointDiffer: https://github.com/apache/ozone/blob/9fe6d106acc71fa79402880ba6344fcab131185d/hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java#L750

**Review by hemantk-12 — COMMENTED** [review]
Changes look good to me. Left few minor comments.
 One general comment, your alignment is bit off. you are using 8 spaces when breaking the line but we use 4 spaces at other places.

**Comment by smengcl** [comment]
> Changes look good to me. Left few minor comments. One general comment, your alignment is bit off. you are using 8 spaces when breaking the line but we use 4 spaces at other places.

Good catch. Looks like a recent IntelliJ upgrade reverted my intentation configs.

I will probably fix those in an addendum.

File: hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java [hunk 5]
```diff
@@ -710,13 +710,64 @@ public synchronized void loadAllCompactionLogs() {
     }
   }
 
+  /**
+   * Helper function that prepends SST file name with SST backup directory path,
+   * and appends the extension '.sst'.
+   */
+  private String getSSTFullPathInBackupDir(String sstFilenameWithoutExtension,
```
[the remark this answers is not in the document]
Comment (smengcl) [reply hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java:717]: Good catch. I forgot to fix the javadoc and method name.

**Pull request merged**